### ISSUE TITLE
Bugs because of NameCleaner

### ISSUE THREAD

**Issue by hrkrshnn** [ethereum/solidity#10428]
Issues caused because of https://github.com/ethereum/solidity/pull/10419 (Note the issues are fixed in that PR, but exists in develop without that.)

```yul
  {
	let a_ := 1
	a_ := 2
	sstore(a_, a_)  
  }
```
`solc --strict-assembly --optimize`
(even on develop) causes

```
: Warning: Yul is still experimental. Please use the output with care.
: Exception in assembler: /solidity/libyul/optimiser/SSATransform.cpp(325): Throw in function virtual void {anonymous}::PropagateValues::operator()(solidity::yul::VariableDeclaration&)
: Dynamic exception type: boost::wrapexcept<solidity::yul::YulAssertion>
: std::exception::what: 
: [solidity::util::tag_comment*] = 
```

**Comment by hrkrshnn** [comment]
```yul  
{
  function f(a_, b) -> c {
    c := a_
    sstore(1, 2)
    sstore(2, 3)
  }
}
```

When going through `--yul-optimizations="Np"` (NameSimplifier and UnusedFunctionParameterPruner) will produce the following code:

```
        { }
        function f(a) -> c
        {
            c := a
            sstore(1, 2)
            sstore(2, 3)
        }
        function f_1(a, b_2) -> c_3
        { c_3 := f(a) }
```

The property that code is disambiguated is false, the variable `a` is repeated. (Note that `libyul/optimiser/Suite.cpp` is modified so that the extra steps at the end are not carried out, stack compressor, VarNameCleaner etc; Normally VarNameCleaner would already rename variables and invalidates the uniqueness property.)

**Comment by hrkrshnn** [comment]
Similar to the last one, but for inliner. `--yul-optimization="Ni"` (NameSimplifier and FullInliner.)
```yul
  {
	let x := f(1)
	sstore(x, x)
	function f(a) -> ret {
	   let c_ := a 
	   ret := add(c_, 1)
	}
  }
```

Produces the non disambiguated
```yul
    {
        {
            let a_1 := 1
            let ret_2 := 0
            let c := a_1
            ret_2 := add(c, 1)
            let x := ret_2
            sstore(x, x)
        }
        function f(a) -> ret
        {
            let c := a
            ret := add(c, 1)
        }
    }
```

I wonder if I can trick the expression simplifier to replace `add(c, 1)` inside `function f` to evaluate to `2` because the value of `c` is `1` outside.  But something like "NiTs" is not able to achieve this.

**Comment by hrkrshnn** [comment]
Closing this since neither me nor @ekpyron could find a serious bug.

The bugs mentioned above will be fixed in 0.7.6.

**Comment by ekpyron** [comment]
I hope we haven't missed anything :-).
